readthedocs/doc_builder/environments.py

```diff
@@ -347,48 +433,30 @@ def handle_exception(self, exc_type, exc_value, _):
                 self.failure = exc_value
             return True
 
-    def run(self, *cmd, **kwargs):
-        """Shortcut to run command from environment."""
-        return self.run_command_class(cls=self.command_class, cmd=cmd, **kwargs)
-
-    def run_command_class(self, cls, cmd, **kwargs):
-        """
-        Run command from this environment.
-
-        Use ``cls`` to instantiate a command
-
-        :param warn_only: Don't raise an exception on command failure
-        """
-        warn_only = kwargs.pop('warn_only', False)
-        # Remove PATH from env, and set it to bin_path if it isn't passed in
-        env_path = self.environment.pop('BIN_PATH', None)
-        if 'bin_path' not in kwargs and env_path:
-            kwargs['bin_path'] = env_path
-        assert 'environment' not in kwargs, "environment can't be passed in via commands."
-        kwargs['environment'] = self.environment
-        kwargs['build_env'] = self
-        build_cmd = cls(cmd, **kwargs)
-        self.commands.append(build_cmd)
-        build_cmd.run()
-
-        # Save to database
+    def record_command(self, command):
         if self.record:
-            build_cmd.save()
+            command.save()
+
+    def _log_warn_only(self, msg):
+        # :'(
+        log.warn(LOG_TEMPLATE.format(
+            project=self.project.slug,
+            version=self.version.slug,
```

Comment:
This one can use the version object because the `*BuildEnvironment` are instantiated with this object. No real problem here.